### DISCUSSION TITLE
浅谈数据库并发控制 - 锁和 MVCC · /database-concurrency-control

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#247] [Announcements]
https://draveness.me/database-concurrency-control

**Comment by bloodyparadise** [comment]
对基于时间戳的协议的具体实现有个疑问，这个协议在实现时是否必须利用原子操作或者锁？
首先，多个并发事务对同一条记录的读/写两个时间戳的读写操作也存在冲突问题；
其次，当一个事务A想要成功对某记录执行写操作时，要经历：1.比较读时间戳 2.比较写时间戳 3.修改写时间戳 4.修改记录本身四步，这四步必须是原子的，否则就有可能在步骤2和3之间另一个事务B（其时间戳小于A，但是大于记录的写时间戳）也对该记录进行写操作，导致两个事务的执行不是可线性化的。
不知道实际使用了该协议的数据库，在具体实现时是怎么做的呢？

**Comment by draveness** [comment]
> @bloodyparadise
> 对基于时间戳的协议的具体实现有个疑问，这个协议在实现时是否必须利用原子操作或者锁？
> 首先，多个并发事务对同一条记录的读/写两个时间戳的读写操作也存在冲突问题；
> 其次，当一个事务A想要成功对某记录执行写操作时，要经历：1.比较读时间戳 2.比较写时间戳 3.修改写时间戳 4.修改记录本身四步，这四步必须是原子的，否则就有可能在步骤2和3之间另一个事务B（其时间戳小于A，但是大于记录的写时间戳）也对该记录进行写操作，导致两个事务的执行不是可线性化的。
> 不知道实际使用了该协议的数据库，在具体实现时是怎么做的呢？

这确实是一个好问题，PostgreSQL 使用的就是时间戳的方式，它每次进行写操作时都会创建新的数据版本，事务**提交时**才会与原始数据进行比较，如果原始数据没有被其他事务改变，就会写入，否则就不会写入。

另外，协议其实并没有对数据库的具体实现细节做约束，如果真要讨论到实现细节，一定是需要原子操作才能保证不会出现问题，我不是特别清楚 PostgreSQL 是如何实现该协议的，可能没办法回答你的这部分问题。

**Comment by bloodyparadise** [comment]
> @Draveness
> > @bloodyparadise
> > 对基于时间戳的协议的具体实现有个疑问，这个协议在实现时是否必须利用原子操作或者锁？
> > 首先，多个并发事务对同一条记录的读/写两个时间戳的读写操作也存在冲突问题；
> > 其次，当一个事务A想要成功对某记录执行写操作时，要经历：1.比较读时间戳 2.比较写时间戳 3.修改写时间戳 4.修改记录本身四步，这四步必须是原子的，否则就有可能在步骤2和3之间另一个事务B（其时间戳小于A，但是大于记录的写时间戳）也对该记录进行写操作，导致两个事务的执行不是可线性化的。
> > 不知道实际使用了该协议的数据库，在具体实现时是怎么做的呢？
> 
> 这确实是一个好问题，PostgreSQL 使用的就是时间戳的方式，它每次进行写操作时都会创建新的数据版本，事务**提交时**才会与原始数据进行比较，如果原始数据没有被其他事务改变，就会写入，否则就不会写入。
> 
> 另外，协议其实并没有对数据库的具体实现细节做约束，如果真要讨论到实现细节，一定是需要原子操作才能保证不会出现问题，我不是特别清楚 PostgreSQL 是如何实现该协议的，可能没办法回答你的这部分问题。
>   

感谢解答: )

**Comment by footmanFF** [comment]
最近也在研究这块东西，《数据库系统概念》介绍了MVCC和两阶段锁的详细的运行机制

**Comment by YoleYu** [comment]
请问画图工具用的是什么啊？这个风格很好看

**Comment by hello2333** [comment]
> MySQL 中实现的多版本两阶段锁协议（Multiversion 2PL）将 MVCC 和 2PL 的优点结合了起来，每一个版本的数据行都具有一个唯一的时间戳，当有读事务请求时，数据库程序会直接从多个版本的数据项中具有最大时间戳的返回。

按照这样的描述，可重复读是如何实现的呢？
假设事务A读取一条记录；事务B对这些记录做修改，然后提交；事务A再读取一次这条记录，如果每次都将“最大时间戳”返回，是不是就违背可重复读了？

**Comment by hello2333** [comment]
> "MySQL 与 MVCC"部分：更新操作就稍微有些复杂了，事务会先读取最新版本的数据计算出数据更新后的结果，然后创建一个新版本的数据，新数据的时间戳是目前数据行的最大版本 ＋1

按照这样说，一行数据会在MySQL中存储多个版本吗？这样岂不是很浪费？

**Comment by OrdinaryYZH** [comment]
请问博主，能否推荐些数据库方面的书籍？

**Comment by draveness** [comment]
> @genericyzh
> 请问博主，能否推荐些数据库方面的书籍？

《数据库系统概念》、《数据库索引设计与优化》

**Comment by OrdinaryYZH** [comment]
对这段话有点不太理解：
> 当我们拥有了不同粒度的锁之后，如果某个事务想要锁定整个数据库或者整张表时只需要简单的锁住对应的节点就会在当前节点加上显示（explicit）锁，在所有的子节点上加隐式（implicit）锁；虽然这种不同粒度的锁能够解决父节点被加锁时，子节点不能被加锁的问题

e.g. trans1: 对表A加S锁了，A中的行都加上了隐式锁；
      trans2: 读取A表中id=1的行，这时候id=1的行不加S锁？
这个处理过程是如何呢？
还有“子节点不能被加锁”作用是什么呢？隐式锁跟X、S锁有啥关系？

**Comment by gblau** [comment]
哇。求画图工具

**Comment by 0xtonyxia** [comment]
那个加入了IS、IX的锁兼容表格，需要说明，S、X是表级锁，行级的S、X是可以和IS、IX兼容的。

**Comment by TennyZhuang** [comment]
这个图是用什么工具画的？

**Comment by MrTreasure** [comment]
请问图片是用什么工具画的？

**Comment by JumpJumpSparrow** [comment]
作者在文章末尾说了> @YoleYu
> 请问画图工具用的是什么啊？这个风格很好看

**Comment by gaotuan** [comment]
> @hello2333
> > "MySQL 与 MVCC"部分：更新操作就稍微有些复杂了，事务会先读取最新版本的数据计算出数据更新后的结果，然后创建一个新版本的数据，新数据的时间戳是目前数据行的最大版本 ＋1
> 
> 按照这样说，一行数据会在MySQL中存储多个版本吗？这样岂不是很浪费？

老版本的数据会在没有回滚段引用是purge掉

**Comment by gaotuan** [comment]
> @hello2333
> > MySQL 中实现的多版本两阶段锁协议（Multiversion 2PL）将 MVCC 和 2PL 的优点结合了起来，每一个版本的数据行都具有一个唯一的时间戳，当有读事务请求时，数据库程序会直接从多个版本的数据项中具有最大时间戳的返回。
 
按照这样的描述，可重复读是如何实现的呢？
假设事务A读取一条记录；事务B对这些记录做修改，然后提交；事务A再读取一次这条记录，如果每次都将“最大时间戳”返回，是不是就违背可重复读了？
这个要看事务的隔离级别，
如果是rc，那么会读到B事务的修改
如果是rr，那么不会读到B事务的修改，在这种情况下会对比这个记录的最新事务id，发现比当前事务id大，回利用undo构造数据的前镜像

**Comment by yuanjize** [comment]
> @OrdinaryYZH
> 对这段话有点不太理解：
> > 当我们拥有了不同粒度的锁之后，如果某个事务想要锁定整个数据库或者整张表时只需要简单的锁住对应的节点就会在当前节点加上显示（explicit）锁，在所有的子节点上加隐式（implicit）锁；虽然这种不同粒度的锁能够解决父节点被加锁时，子节点不能被加锁的问题
> 
> e.g. trans1: 对表A加S锁了，A中的行都加上了隐式锁；
>       trans2: 读取A表中id=1的行，这时候id=1的行不加S锁？
> 这个处理过程是如何呢？
> 还有“子节点不能被加锁”作用是什么呢？隐式锁跟X、S锁有啥关系？

t1给表A中的一条数据a加了S锁，这个时候t2想给A加X表锁，这个时候如果没有意向锁，给A加成功了那么相当于给A下面的每条数据也加上了X锁。这个时候t1和t2同时给记录a分别加上了s和x锁，这明显有问题了。
如果有意向锁 t2发现表A上有IS，那么加X锁就会失败了。

**Comment by ricemaster** [comment]
您好，google mvcc时看到了这篇文章，收获颇多，非常感谢分享。

看完有一个小问题，请问前面讲了
> 锁机制和基于时间戳排序的协议是悲观的

那为什么pg中的多版本时间戳排序协议（Multiversion Timestamp Ordering）就算作了MVCC+乐观的结合了呢？
